### PR TITLE
Form groups: label + input + helper text

### LINKED ISSUE #413: Error text for form elements
It would be nice to have additional UI items for presenting form errors:

- A font style that's smaller (eg caption) that can be used to do small captions under a form element in a vertical stacked presentation
- A style class for form error text (non-error would just use something like `pt-text-muted` for captions)

### PR DESCRIPTION
**Pull request by leebyp**
#### Checklist

- [x] Update documentation

@llorca @thisisalessandro
#### Reviewers should focus on:

- Wording in documentation - is this enough to show the variants of controls that we provide?
- Class naming

screen shot 2017-02-01 at 11 35 19

screen shot 2017-02-01 at 12 01 29

with `.pt-inline`

screen shot 2017-02-01 at 11 37 08

screen shot 2017-02-01 at 12 01 33

### CODE REVIEW

File: packages/core/src/components/forms/_helper-text.scss [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        color: $color;
+      }
+    }
+  }
+
+  &.pt-inline {
+    flex-direction: row;
+
+    > label {
+      margin: ($pt-grid-size * 0.7) $pt-grid-size 0 0;
```
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: This makes me sad, what's a good way to do this?

**Comment by thisisalessandro** [comment]
@llorca should intents automatically add icons here?

**Comment by leebyp** [comment]
@thisisalessandro sorry that's my bad, i put the class on the wrong element in the screenshot: updated

**Comment by thisisalessandro** [comment]
I think whatever icons we have inside should be left unchanged. Let me rephrase my question: should adding `.pt-intent-danger` add a `pt-icon-error` icon inside the input field, on the right? This may be very Palantir-specific because it's part of the guidelines, although it's a good general accessibility practice to always pair a color with an icon (so as not to rely solely on one of the two).

**Comment by llorca** [comment]
@thisisalessandro Yes, but very Palantir-specific. This is part of the "delta" that we need to address after we integrate the guidelines. Nothing that needs to be added to this PR.

The icon are now properly colored when there's an intent, so we're good on that front 👍

File: packages/core/src/components/forms/_helper-text.scss [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      <span class="pt-icon pt-icon-calendar"></span>
+      <input id="example-form-group-input" class="pt-input" style="width: 200px;" type="text" placeholder="Input group" dir="auto" />
+    </div>
+    <div class="pt-helper-text">Helper text with details / user feedback</div>
+  </div>
+</div>
+
+.pt-inline - Inline
+
+Weight: 1.1
```
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: should i be reordering the rest of them to keep integers?

File: packages/core/src/components/forms/_helper-text.scss [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+- Add helper text to the form group to reflect user input (display intent with `.pt-helper-text.pt-intent-*`).
+
+Markup:
+<div class="pt-form-group {{.modifier}}">
+  <label for="example-form-group-input">
+    Label A
+    <span class="pt-text-muted">(optional)</span>
+  </label>
+  <div class="form-content">
```
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: i was thinking about that - we don't have any css for it, but i do agree with keeping everything prefixed

File: packages/docs/src/styles/_sections.scss [hunk 4]
```diff
@@ -69,6 +69,7 @@
 }
 
 #{section-name("Labels")},
+#{section-name("Helper Text")},
```
Comment (leebyp) [inline packages/docs/src/styles/_sections.scss]: using `line-height` breaks down as soon as the label overflows to a second line (or do we want to leave it to the consumer on how to handle? (whether it be with ellipsis etc?)

Comment (llorca) [inline packages/docs/src/styles/_sections.scss]: Totally right! Then I guess your `margin` solution sounds good then. Instead of `$pt-grid-size * 0.7`, would it be possible to compute it? Something like `($pt-input-height - $label-font-size) / 2`? Same for large? @giladgray will know better

File: packages/core/src/components/forms/_helper-text.scss (hunk 2 above)
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: i didn't even know you could do non-integer weights!

File: packages/core/src/components/forms/_helper-text.scss [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+// of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+// and https://github.com/palantir/blueprint/blob/master/PATENTS
+
+@import "../../common/variables";
+@import "../../common/mixins";
+
+/*
+Helper Text
+
+Add additional helper text to communicate user input. Additional markup is required for the more
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: nailed it on both counts

File: packages/core/src/common/classes.ts [hunk 6]
```diff
@@ -83,6 +83,8 @@ export const INTENT_WARNING = "pt-intent-warning";
 export const INTENT_DANGER = "pt-intent-danger";
 
 export const LABEL = "pt-label";
+export const FORM_GROUP = "pt-form-group";
+export const HELPER_TEXT = "pt-helper-text";
```
Comment (leebyp) [inline packages/core/src/common/classes.ts]: `FORM-HELPER-TEXT`? prepending with `pt-form-*` reflects the usage

File: packages/docs/src/styles/_sections.scss (hunk 4 above)
Comment (leebyp) [inline packages/docs/src/styles/_sections.scss]: on second thought, im going to stick with `line-height` since that's what we did previously for simple labels too - if we want to change it, we should do so together in a later PR

File: packages/core/src/components/forms/_helper-text.scss [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  <div class="pt-form-content">
+    <div class="pt-input-group">
+      <span class="pt-icon pt-icon-calendar"></span>
+      <input id="example-form-group-input" class="pt-input" style="width: 200px;" type="text" placeholder="Input group" dir="auto" />
+    </div>
+    <div class="pt-form-helper-text">Helper text with details / user feedback</div>
+  </div>
+</div>
+
+.pt-inline - Inline
```
Comment (llorca) [inline packages/core/src/components/forms/_helper-text.scss]: "Inline label"

Maybe let's add green/red intent for the example?

**Comment by llorca** [comment]
This is sweet! In addition, what do you think about supporting these:
- `[:required](https://developer.mozilla.org/en-US/docs/Web/CSS/:required)` to show a `(required)` near the label
- `[:optional](https://developer.mozilla.org/en-US/docs/Web/CSS/:optional)` to show a `(optional)` near the label
- `[:valid](https://developer.mozilla.org/en-US/docs/Web/CSS/:valid)` to extend `.pt-intent-success` styles
- `[:invalid](https://developer.mozilla.org/en-US/docs/Web/CSS/:invalid)` to extend `.pt-intent-danger` styles

File: packages/core/src/components/forms/_helper-text.scss (hunk 7 above)
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: changed to just use the markup, showing the variants - the `modifier` table doesn't work here

File: packages/core/src/components/forms/_helper-text.scss [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  Note that this feature necessitates additional DOM elements,
+  thus requiring a more complex form structure achieved with `.pt-form-group`.
+</div>
+
+- Link each label to its respective control element with a `for={#id}` attribute on the `<label>` and
+`id={#id}` on the control.
+
+- Add extra information to the label with `span.pt-text-muted`.
+
+- Add helper text to the form group to reflect user input (display intent with `.pt-form-helper-text.pt-intent-*`).
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: "Add `.pt-form-helper-text` ..." (you use class name in all the other bullets).

File: packages/core/src/components/forms/_helper-text.scss [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+.pt-form-group {
+  display: flex;
+  flex-direction: column;
+  margin: 0 0 $pt-grid-size;
+
+  > label {
+    margin-bottom: $pt-grid-size / 2;
+  }
+
+  > .pt-form-content > .pt-form-helper-text {
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: do we _need_ `>`s here? would it still work without them? my suspicion is yes.

is it possible to nest a `.pt-form-group` inside another `.pt-form-group`? my suspicion is no. if so, that'd be the reason for child selectors.

File: packages/core/src/components/forms/_helper-text.scss [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    <div class="pt-form-helper-text">Helper text with details / user feedback</div>
+  </div>
+</div>
+
+Weight: 1.1
+
+Styleguide components.forms.helper-text
+*/
+
+.pt-form-group {
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: hmm seems like this file should be called `_form-group.scss`

**Comment by giladgray** [comment]
re @llorca's comment above: `:required`, `:optional` cannot add elements because inputs are replaced and can't have `::before/after`. form validation can be touchy; users may want to handle it themselves in code and apply intent classes rather than relying on built-in validators.

File: packages/core/src/components/forms/_helper-text.scss [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+- Link each label to its respective control element with a `for={#id}` attribute on the `<label>` and
+`id={#id}` on the control.
+
+- Add extra information to the label with `span.pt-text-muted`.
+
+- Add helper text to the form group to reflect user input (display intent with `.pt-form-helper-text.pt-intent-*`).
+
+- Add `.pt-inline` to `.pt-form-group` to place the label to the left of the control.
+
+- Add `.pt-large` to `.pt-form-group` to align the label when used with large Blueprint controls.
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: note about also adding `.pt-large` to input group etc?

File: packages/core/src/components/forms/_helper-text.scss [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+  </div>
+</div>
+<div class="pt-form-group">
+  <label for="example-form-group-input-b">
+    Label B
+    <span class="pt-text-muted">(required)</span>
+  </label>
+  <div class="pt-form-content">
+    <input id="example-form-group-input-b" class="pt-input pt-intent-danger" style="width: 300px;" type="text" dir="auto" />
+    <div class="pt-form-helper-text pt-intent-danger">Label B is required</div>
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: this text is redundant with the label. how about: "Please enter a value."

File: packages/core/src/components/forms/_helper-text.scss [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    <div class="pt-form-helper-text">Helper text with details / user feedback</div>
+  </div>
+</div>
+<div class="pt-form-group">
+  <label for="example-form-group-input-b">
+    Label B
+    <span class="pt-text-muted">(required)</span>
+  </label>
+  <div class="pt-form-content">
+    <input id="example-form-group-input-b" class="pt-input pt-intent-danger" style="width: 300px;" type="text" dir="auto" />
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: placeholder?

File: packages/core/src/components/forms/_helper-text.scss [hunk 14]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+// Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+// of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+// and https://github.com/palantir/blueprint/blob/master/PATENTS
+
+@import "../../common/variables";
+@import "../../common/mixins";
+
+/*
+Helper text
```
Comment (giladgray) [inline packages/core/src/components/forms/_helper-text.scss]: i think this is actually about "Form groups," which happen to support helper text.

File: packages/core/src/components/forms/_helper-text.scss (hunk 8 above)
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: this meant a new element - updated to say that

File: packages/core/src/components/forms/_helper-text.scss (hunk 11 above)
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: `large Blueprint controls` isn't enough to convey?

File: packages/core/src/components/forms/_helper-text.scss (hunk 13 above)
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: @thisisalessandro i copied the example from the guidelines so you'll wanna update too

File: packages/core/src/components/forms/_helper-text.scss (hunk 9 above)
Comment (leebyp) [inline packages/core/src/components/forms/_helper-text.scss]: It will still work, but we're specifying a pretty specific DOM structure here for the `form-group`, hence the `>`.

I can see nested form groups, but if we don't want to support, not strongly inclined.

File: packages/core/src/components/forms/_form-group.scss [hunk 15]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+// Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+// of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+// and https://github.com/palantir/blueprint/blob/master/PATENTS
+
+@import "../../common/variables";
+@import "../../common/mixins";
+
+/*
+Form group
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: Form group**s**

File: packages/core/src/components/forms/_form-group.scss [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+Add additional helper text to help users navigate forms and complex fields.
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: make this about Form group as a "power label." also remember #600 when describing the component.

File: packages/core/src/components/forms/_form-group.scss [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+<div class="pt-callout pt-intent-primary pt-icon-info-sign">
+  Note that this feature necessitates additional DOM elements,
+  thus requiring a more complex form structure achieved with `.pt-form-group`.
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: this comment lacks context: more complex than what? 
a more complete explanation may obviate the need for this callout.

File: packages/core/src/components/forms/_form-group.scss [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+- Add an element with `.pt-form-helper-text` to reflect user input (display intent with `.pt-intent-*`).
+
+- Add `.pt-inline` to `.pt-form-group` to place the label to the left of the control.
+
+- Add `.pt-large` to `.pt-form-group` to align the label when used with large Blueprint controls.
+
+Markup:
+<div class="pt-form-group">
+  <label for="example-form-group-input-a">
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: API question: why do these labels not use `.pt-label`?

this section in the docs immediately follows "Labels" where we seem to instruct users to write `label.pt-label`, but suddenly we're not following our own rules.

**Comment by bsr203** [comment]
hi. instead of setting the classes to both `input`, `help message`, ... or whatever contained element within field group, is it easier to use a class `has-error` at field group level?  Similar to http://v4-alpha.getbootstrap.com/components/forms/#how-it-works . This may add some restriction on the structure of elements within field group, and need additional styling but only need to set the class at top level.

**Comment by leebyp** [comment]
@bsr203 good point, similar to `InputGroup`, I've allowed `.pt-disabled`, `.pt-intent-*` as classes on `.pt-form-group`. By the same reasoning, the controls themselves need to be styled separately - and should be tackled in a component if we decide to build it.

File: packages/core/src/components/forms/_form-group.scss [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+  &.pt-inline {
+    flex-direction: row;
+
+    label.pt-label {
+      margin: 0 $pt-grid-size 0 0;
+      line-height: $pt-input-height;
+    }
+  }
+
+  &.pt-large label.pt-label {
```
Comment (llorca) [inline packages/core/src/components/forms/_form-group.scss]: This rule is only accurate for the `pt-inline` case. If you're not inline, the line-height shouldn't change, otherwise we get inconsistent spacing between labels and inputs in the `input.pt-large` case and `.pt-form-group.pt-large` case

File: packages/core/src/components/forms/_label.scss [hunk 20]
```diff
@@ -8,19 +8,22 @@
...
-Add labels to inputs to enhance the usability of your forms.
+Labels enhance the usability of your forms.
 
-Putting the `<input>` element _inside_ a `<label>` element also increases the area where the user
-can click to activate the control. Notice how in the example below, clicking a `<label>` below
-focuses its `<input>`.
+- Add extra information to the label with `span.pt-label-description`.
+
+- Putting the `<input>` element _inside_ a `<label>` element increases the area where the user
+can click to activate the control. Notice how in the examples below, clicking a `<label>` focuses its `<input>`.
```
Comment (llorca) [inline packages/core/src/components/forms/_label.scss]: What about saying that `input` inside `label` is the "deprecated way" of buildings form, and mentioning that you won't be able to use for complex form field (i.e. control groups, numeric inputs...). In addition, we could say that this method of putting `input` inside `label` is an alternative way of binding label to field (`for` + `id` is the new way we recommend).

Comment (leebyp) [inline packages/core/src/components/forms/_label.scss]: If that's the way we're going, how about also explicitly putting that this **will** be deprecated in 2.0?

Comment (llorca) [inline packages/core/src/components/forms/_label.scss]: We actually need to make a decision whether or not we want to deprecate the older way of writing forms (`input` inside `label`). Once we decide, we can put the appropriate wording in the docs 👍 

@giladgray @adidahiya @cmslewis See above, need your thoughts. Do we:
1. Deprecate the older way of writing forms (`input` inside `label`)
1. Keep the older way (`input` inside `label`) alongside the new way (`pt-form-group` + `for` + `id`) and maintain these two approaches over time.

I see a valid use case for keeping both around: `input` inside `label` can be used for dead simple form fields, where `pt-form-group` should be used for more complex ones (i.e. ones that require control groups for instance.)

I can also see a valid reason to remove the super simple way: it doesn't support all the use cases (no way to make control group work cleanly inside a `label`), and it's easier to maintain one approach.

Comment (cmslewis) [inline packages/core/src/components/forms/_label.scss]: Frankly, my dream scenario is to find a way to support arbitrary form elements (`pt-input`, `pt-control-group`, `pt-button-group`, etc.) inside of a `label`. From a user perspective, I'd expect at least that common pattern to work, and I'd be thrown off by the current behavior of `pt-control-group` in that context.

Barring that, the best approach in my mind is to keep `label` around as a "leaf" element (one that can't have children) and require use of `pt-form-group` for the `for`-with-`id` use case.

So I guess I'm arguing for No. 1 above.

File: packages/core/src/components/forms/_form-group.scss [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+
+<div class="pt-callout pt-intent-primary pt-icon-info-sign">
+  Note that this feature necessitates additional DOM elements,
+  thus requiring a more complex form structure achieved with `.pt-form-group`.
+</div>
+
+- Link each label to its respective control element with a `for={#id}` attribute on the `<label>` and
+`id={#id}` on the control.
+
+- Add extra information to the label with `span.pt-text-muted`.
```
Comment (llorca) [inline packages/core/src/components/forms/_form-group.scss]: I personally think the block of examples with markup speaks for itself already

File: packages/core/src/components/forms/_form-group.scss [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+  </label>
+  <div class="pt-form-content">
+    <input id="example-form-group-input-b" class="pt-input pt-intent-danger" style="width: 300px;" placeholder="Input text" type="text" dir="auto" />
+    <div class="pt-form-helper-text">Please enter a value</div>
+  </div>
+</div>
+<div class="pt-form-group pt-inline pt-large pt-disabled">
+  <label class="pt-label" for="example-form-group-input-c">
+    Label C
+    <span class="pt-label-description">(required)</span>
```
Comment (llorca) [inline packages/core/src/components/forms/_form-group.scss]: `(optional)` for the sake of the example

File: packages/core/src/components/forms/_form-group.scss [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+    <div class="pt-form-helper-text">Helper text with details / user feedback</div>
+  </div>
+</div>
+<div class="pt-form-group pt-intent-danger">
+  <label class="pt-label" for="example-form-group-input-b">
+    Label B
+    <span class="pt-label-description">(required)</span>
+  </label>
+  <div class="pt-form-content">
+    <input id="example-form-group-input-b" class="pt-input pt-intent-danger" style="width: 300px;" placeholder="Input text" type="text" dir="auto" />
```
Comment (llorca) [inline packages/core/src/components/forms/_form-group.scss]: `placeholder="Placeholder text"`

File: packages/core/src/components/forms/_form-group.scss [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+Markup:
+<div class="pt-form-group">
+  <label class="pt-label" for="example-form-group-input-a">
+    Label A
+    <span class="pt-label-description">(required)</span>
+  </label>
+  <div class="pt-form-content">
+    <div class="pt-input-group">
+      <span class="pt-icon pt-icon-calendar"></span>
+      <input id="example-form-group-input-a" class="pt-input" style="width: 200px;" type="text" placeholder="Input group" dir="auto" />
```
Comment (llorca) [inline packages/core/src/components/forms/_form-group.scss]: `placeholder="Placeholder text"`

File: packages/core/src/components/forms/_form-group.scss [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+
+Markup:
+<div class="pt-form-group">
+  <label class="pt-label" for="example-form-group-input-a">
+    Label A
+    <span class="pt-label-description">(required)</span>
+  </label>
+  <div class="pt-form-content">
+    <div class="pt-input-group">
+      <span class="pt-icon pt-icon-calendar"></span>
```
Comment (llorca) [inline packages/core/src/components/forms/_form-group.scss]: I'd put the icon in the next example so we can see how it renders with an intent

File: packages/core/src/components/forms/_form-group.scss (hunk 21 above)
Comment (leebyp) [inline packages/core/src/components/forms/_form-group.scss]: actually this should be possible with simple labels too so i've moved it up to that section (and needs a class to help with the disabled state)

File: packages/core/src/components/forms/_form-group.scss [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+Styleguide components.forms.form-group
+*/
+
+.pt-form-group {
+  display: flex;
+  flex-direction: column;
+  margin: 0 0 $pt-grid-size;
+
+  > label {
```
Comment (leebyp) [inline packages/core/src/components/forms/_form-group.scss]: updated to use `label.pt-label` to mimic the css in the simple version

**Comment by leebyp** [comment]
@llorca @cmslewis Updated the docs with references between Labels and Form groups. I don't think we should deprecate the css logic yet until v2.0 since that's a breaking change.

**Comment by llorca** [comment]
If we decide to deprecate, we can (should) do it now (i.e. add a callout that says it will be removed in 2.0). Deprecation doesn't mean we're removing it now, instead we're warning users that it'll be removed in the next major version, so they have time to adjust

**Comment by leebyp** [comment]
Ah gotcha, I'm becoming more inclined to keep both (even if it means we have to support that), because of the simplicity in the original structure. `for={...} id={...}` is hard to keep track of.

I can't find my original comment anymore but what were thoughts on reordering the docs structure to be:
- Labels
    - Simple labels
    - Form groups
    - Disabled labels

**Comment by cmslewis** [comment]
@leebyp 

>
> - Labels
>   - Simple labels
>   - Form groups
>   - Disabled labels

^ I'm fine with this.

And I'd prefer to keep both around as long as we're clear how `pt-label`s should be used.

File: packages/core/src/components/forms/_label.scss [hunk 27]
```diff
@@ -111,14 +135,24 @@ label.pt-label {
...
     color: $pt-dark-heading-color;
 
+    .pt-label-description {
+      color: $pt-text-color-muted;
+    }
+
     &.pt-disabled {
-      color: $pt-dark-text-color-disabled;
+      &,
+      .pt-label-description {
```
Comment (llorca) [inline packages/core/src/components/forms/_label.scss]: `.pt-text-muted { color: $pt-text-color-disabled; }` isn't that nice either. Is there an elegant way to get rid of `.pt-label-description`?

File: packages/core/src/components/forms/_label.scss [hunk 28]
```diff
@@ -136,21 +139,21 @@ label.pt-label {
...
 
-    .pt-label-description {
+    .pt-muted {
       color: $pt-text-color-muted;
     }
 
     &.pt-disabled {
       &,
-      .pt-label-description {
+      .pt-muted {
```
Comment (giladgray) [inline packages/core/src/components/forms/_label.scss]: wait what's `pt-muted`? this is also a new modifier.

**did you mean `pt-text-muted`?**

if you did mean `pt-text-muted` then you don't need to set the original color, only the disabled color.

File: packages/core/src/components/forms/_label.scss [hunk 29]
```diff
@@ -8,19 +8,41 @@
...
+Labels enhance the usability of your forms.
 
-Putting the `<input>` element _inside_ a `<label>` element also increases the area where the user
-can click to activate the control. Notice how in the example below, clicking a `<label>` below
-focuses its `<input>`.
+<div class="pt-callout pt-intent-success pt-icon-comparison">
+  <h5>Simple labels vs. form groups</h5>
+  <p>Blueprint provides two ways of connecting label text to control fields, depending on the complexity of the control.</p>
+  <p>Simple labels are a basic way to connect a label with a single control.</p>
+  <p>Form groups require more markup to maintain consistent visuals across more complex controls.</p>
```
Comment (giladgray) [inline packages/core/src/components/forms/_label.scss]: put the reason first:

"Form groups support more complex control layouts but require more markup to maintain consistent visuals."

File: packages/core/src/components/forms/_label.scss [hunk 30]
```diff
@@ -8,19 +8,41 @@
...
+
+Weight: 1
+
+Styleguide components.forms.labels
+*/
+
+/*
+Simple labels
+
+Simple labels are useful for basic forms eg. for a single `<input>`. This structure is not sufficient for some complex form fields such as
```
Comment (giladgray) [inline packages/core/src/components/forms/_label.scss]: remove "eg." - not used correctly. 

"Simple labels are useful for basic forms for a single `<input>`."

File: packages/core/src/components/forms/_label.scss [hunk 31]
```diff
@@ -8,19 +8,41 @@
...
+[numeric inputs](#components.forms.numeric-input) or [control groups](#components.forms.control-group). These require more complicated
+[form groups](#components.forms.labels.form-groups).
```
Comment (giladgray) [inline packages/core/src/components/forms/_label.scss]: i don't think these last two sentences are helpful here cuz they're repeated elsewhere in the docs. please remove so it's just about simple labels.

File: packages/core/src/components/forms/_form-group.scss [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+// of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+// and https://github.com/palantir/blueprint/blob/master/PATENTS
+
+@import "../../common/variables";
+@import "../../common/mixins";
+
+/*
+Form groups
+
+Form groups allow for more complex forms and fields than [simple labels](#components.forms.labels.simple-labels), for example,
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: "Form groups support more complex form controls than [simple labels], such as [control groups] or [`NumericInput`s]. They also support additional helper text to aid with user navigation."

File: packages/core/src/components/forms/_form-group.scss [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+with [numeric inputs](#components.forms.numeric-input) or [control groups](#components.forms.control-group).
+They also allow the addition of helper text to aid with user navigation.
+
+<div class="pt-callout pt-intent-primary pt-icon-info-sign">
+  Note that this feature necessitates additional DOM elements, thus requiring a more complex structure achieved with `.pt-form-group`.
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: this callout is confusing... "additional DOM elements" compared to what? i think it might be safe to remove.

File: packages/core/src/components/forms/_form-group.scss [hunk 34]
```diff
@@ -0,0 +1,144 @@
...
+</div>
+
+- Link each label to its respective control element with a `for={#id}` attribute on the `<label>` and
+`id={#id}` on the control.
+
+- Add `.pt-intent-*` or `.pt-disabled` to `.pt-form-group` to style the label and helper text. Similar to labels, nested controls need to be styled separately.
+
+- Add `.pt-inline` to `.pt-form-group` to place the label to the left of the control.
+
+- Add `.pt-large` to `.pt-form-group` to align the label when used with large inline Blueprint controls.
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: please wrap all comments to 100 characters if you can. VS Code "rewrap" extension is very useful here.

File: packages/core/src/components/forms/_form-group.scss [hunk 35]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    label.pt-label {
+      margin: 0 $pt-grid-size 0 0;
+      line-height: $pt-input-height;
+    }
+  }
+
+  // pt-disable always overrides over styles
+  // stylelint-disable declaration-no-important
```
Comment (giladgray) [inline packages/core/src/components/forms/_form-group.scss]: use a `stylelint-disable-next-line` and you won't have to re-enable
